src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -41,30 +47,98 @@ public sealed class PeerSelector : IPeerSelector
         /// Constructor for the peer selector.
         /// </summary>
         /// <param name="peerAddresses">The collection of peer address as managed by the peer address manager.</param>
-        public PeerSelector(ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
+        public PeerSelector(ILoggerFactory loggerFactory, ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
         {
-            Guard.NotNull(peerAddresses, nameof(peerAddresses));
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
             this.peerAddresses = peerAddresses;
         }
 
         /// <inheritdoc/>
         public PeerAddress SelectPeer()
         {
-            var tried = this.peerAddresses.Attempted().Concat(this.peerAddresses.Connected());
-            if (tried.Any() == true && (!this.peerAddresses.Fresh().Any() || new Random().Next(2) == 0))
-                return tried.Random();
+            this.logger.LogTrace("()");
 
-            if (this.peerAddresses.Fresh().Any())
-                return this.peerAddresses.Fresh().Random();
+            PeerAddress peerAddress = null;
 
-            return null;
+            var peers = SelectPeers();
+            if (peers.Any())
+            {
+                peerAddress = peers.Random();
+                this.logger.LogTrace("(-):{0}={1}", nameof(peerAddress), peerAddress.NetworkAddress.Endpoint.ToString());
+            }
+            else
+                this.logger.LogTrace("(-):{0} [NULL]", nameof(peerAddress));
```

Comment:
`this.logger.LogTrace("(-)[NO_PEERS]");`

Reply:
Fixed